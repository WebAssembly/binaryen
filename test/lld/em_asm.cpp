#include <emscripten/em_asm.h>

int main() {
  EM_ASM({ Module.print("Hello world"); });
  int x = EM_ASM_INT({ return $0 + $1; }, 13, 27);
  EM_ASM_({ Module.print("Got " + $0); }, x);
  return 0;
}
