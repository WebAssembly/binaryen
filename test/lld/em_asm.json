{ "asmConsts": {"2": ["{ Module.print(\"Got \" + $0); }", ["ii"]],"0": ["{ Module.print(\"Hello world\"); }", ["i"]],"1": ["{ return $0 + $1; }", ["iii"]]},"staticBump": 84, "initializers": [] }
