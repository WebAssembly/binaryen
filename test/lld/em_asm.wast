(module
 (type $0 (func (param i32) (result i32)))
 (type $1 (func (param i32 i32 i32) (result i32)))
 (type $2 (func (param i32 i32) (result i32)))
 (type $3 (func (result i32)))
 (type $4 (func))
 (import "env" "_Z24emscripten_asm_const_intIJEEiPKcDpT_" (func $_Z24emscripten_asm_const_intIJEEiPKcDpT_ (param i32) (result i32)))
 (import "env" "_Z24emscripten_asm_const_intIJiiEEiPKcDpT_" (func $_Z24emscripten_asm_const_intIJiiEEiPKcDpT_ (param i32 i32 i32) (result i32)))
 (import "env" "_Z24emscripten_asm_const_intIJiEEiPKcDpT_" (func $_Z24emscripten_asm_const_intIJiEEiPKcDpT_ (param i32 i32) (result i32)))
 (global $global$0 (mut i32) (i32.const 66192))
 (global $global$1 i32 (i32.const 66192))
 (global $global$2 i32 (i32.const 652))
 (table 1 1 funcref)
 (memory $0 2)
 (data (i32.const 568) "{ Module.print(\"Hello world\"); }\00{ return $0 + $1; }\00{ Module.print(\"Got \" + $0); }\00")
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "main" (func $main))
 (export "__heap_base" (global $global$1))
 (export "__data_end" (global $global$2))
 (func $main (; 3 ;) (type $3) (result i32)
  (drop
   (call $_Z24emscripten_asm_const_intIJEEiPKcDpT_
    (i32.const 568)
   )
  )
  (drop
   (call $_Z24emscripten_asm_const_intIJiEEiPKcDpT_
    (i32.const 621)
    (call $_Z24emscripten_asm_const_intIJiiEEiPKcDpT_
     (i32.const 601)
     (i32.const 13)
     (i32.const 27)
    )
   )
  )
  (i32.const 0)
 )
 (func $__wasm_call_ctors (; 4 ;) (type $4)
 )
 ;; custom section "linking", size 3
)

