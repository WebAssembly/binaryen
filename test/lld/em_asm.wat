(module
 (type $0 (func (param i32 i32 i32) (result i32)))
 (type $1 (func))
 (type $2 (func (result i32)))
 (type $3 (func (param i32 i32) (result i32)))
 (import "env" "emscripten_asm_const_int" (func $emscripten_asm_const_int (param i32 i32 i32) (result i32)))
 (memory $0 2)
 (data (i32.const 568) "{ Module.print(\"Hello world\"); }\00\00{ return $0 + $1; }\00ii\00{ Module.print(\"Got \" + $0); }\00i\00")
 (table $0 1 1 funcref)
 (global $global$0 (mut i32) (i32.const 66208))
 (global $global$1 i32 (i32.const 658))
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "main" (func $main))
 (export "__data_end" (global $global$1))
 (func $__wasm_call_ctors (; 1 ;) (type $1)
 )
 (func $__original_main (; 2 ;) (type $2) (result i32)
  (local $0 i32)
  (global.set $global$0
   (local.tee $0
    (i32.sub
     (global.get $global$0)
     (i32.const 32)
    )
   )
  )
  (drop
   (call $emscripten_asm_const_int
    (i32.const 568)
    (i32.const 601)
    (i32.const 0)
   )
  )
  (i64.store offset=16
   (local.get $0)
   (i64.const 115964117005)
  )
  (i32.store
   (local.get $0)
   (call $emscripten_asm_const_int
    (i32.const 602)
    (i32.const 622)
    (i32.add
     (local.get $0)
     (i32.const 16)
    )
   )
  )
  (drop
   (call $emscripten_asm_const_int
    (i32.const 625)
    (i32.const 656)
    (local.get $0)
   )
  )
  (global.set $global$0
   (i32.add
    (local.get $0)
    (i32.const 32)
   )
  )
  (i32.const 0)
 )
 (func $main (; 3 ;) (type $3) (param $0 i32) (param $1 i32) (result i32)
  (call $__original_main)
 )
 ;; custom section "producers", size 112
)

