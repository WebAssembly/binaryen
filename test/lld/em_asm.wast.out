(module
 (type $0 (func (param i32) (result i32)))
 (type $1 (func (param i32 i32 i32) (result i32)))
 (type $2 (func (param i32 i32) (result i32)))
 (type $3 (func (result i32)))
 (type $4 (func))
 (type $FUNCSIG$ii (func (param i32) (result i32)))
 (type $FUNCSIG$iiii (func (param i32 i32 i32) (result i32)))
 (type $FUNCSIG$iii (func (param i32 i32) (result i32)))
 (import "env" "emscripten_asm_const_i" (func $emscripten_asm_const_i (param i32) (result i32)))
 (import "env" "emscripten_asm_const_iii" (func $emscripten_asm_const_iii (param i32 i32 i32) (result i32)))
 (import "env" "emscripten_asm_const_ii" (func $emscripten_asm_const_ii (param i32 i32) (result i32)))
 (global $global$0 (mut i32) (i32.const 66656))
 (global $global$1 i32 (i32.const 66656))
 (global $global$2 i32 (i32.const 1108))
 (table 1 1 anyfunc)
 (memory $0 2)
 (data (i32.const 1024) "{ Module.print(\"Hello world\"); }\00{ return $0 + $1; }\00{ Module.print(\"Got \" + $0); }\00")
 (export "memory" (memory $0))
 (export "__wasm_call_ctors" (func $__wasm_call_ctors))
 (export "main" (func $main))
 (export "__heap_base" (global $global$1))
 (export "__data_end" (global $global$2))
 (export "stackSave" (func $stackSave))
 (export "stackAlloc" (func $stackAlloc))
 (export "stackRestore" (func $stackRestore))
 (export "__growWasmMemory" (func $__growWasmMemory))
 (func $main (; 3 ;) (type $3) (result i32)
  (drop
   (call $emscripten_asm_const_i
    (i32.const 0)
   )
  )
  (drop
   (call $emscripten_asm_const_ii
    (i32.const 2)
    (call $emscripten_asm_const_iii
     (i32.const 1)
     (i32.const 13)
     (i32.const 27)
    )
   )
  )
  (i32.const 0)
 )
 (func $__wasm_call_ctors (; 4 ;) (type $4)
  (nop)
 )
 (func $stackSave (; 5 ;) (result i32)
  (get_global $global$0)
 )
 (func $stackAlloc (; 6 ;) (param $0 i32) (result i32)
  (local $1 i32)
  (set_global $global$0
   (tee_local $1
    (i32.and
     (i32.sub
      (get_global $global$0)
      (get_local $0)
     )
     (i32.const -16)
    )
   )
  )
  (get_local $1)
 )
 (func $stackRestore (; 7 ;) (param $0 i32)
  (set_global $global$0
   (get_local $0)
  )
 )
 (func $__growWasmMemory (; 8 ;) (param $newSize i32) (result i32)
  (grow_memory
   (get_local $newSize)
  )
 )
)
;; METADATA: { "asmConsts": {"2": ["{ Module.print(\"Got \" + $0); }", ["ii"]],"0": ["{ Module.print(\"Hello world\"); }", ["i"]],"1": ["{ return $0 + $1; }", ["iii"]]},"staticBump": 84, "initializers": ["__wasm_call_ctors"] }
